earth2studio/utils/interp.py

```diff
@@ -0,0 +1,170 @@
+        * (1 - normed_lon_distance)
+    )
+    result += (
+        values[..., latinds, loninds + 1]
+        * (1 - normed_lat_distance)
+        * (normed_lon_distance)
+    )
+    result += (
+        values[..., latinds + 1, loninds]
+        * (normed_lat_distance)
+        * (1 - normed_lon_distance)
+    )
+    result += (
+        values[..., latinds + 1, loninds + 1]
+        * (normed_lat_distance)
+        * (normed_lon_distance)
+    )
+    return result.reshape(*values.shape[:-2], latshape, lonshape)
+
+
+class LatLonInterpolation(nn.Module):
+    """Bilinear interpolation between arbitrary grids.
+
+    The mapped values can be on arbitrary grid, but the output grid should be
+    contained within the input grid.
+
+    Initializing the interpolation object can be somewhat slow, but interpolation is
+    fast and runs on the GPU once initialized. Therefore, prefer to reuse the
+    interpolation object when possible.
+
+    Parameters
+    ----------
+    lat0 : torch.Tensor [H_in, W_in]
+        Tensor of input latitude coordinates
+    lon0 : torch.Tensor [H_in, W_in]
+        Tensor of input longitude coordinates
+    lat1 : torch.Tensor [W_out, H_out]
+        Tensor of output latitude coordinates
+    lon1 : torch.Tensor [W_out, H_out]
+        Tensor of output longitude coordinates
+    """
+
+    def __init__(
+        self,
+        lat_in: Tensor | ArrayLike,
+        lon_in: Tensor | ArrayLike,
+        lat_out: Tensor | ArrayLike,
+        lon_out: Tensor | ArrayLike,
+    ):
+        super().__init__()
+
+        lat_in = (
+            lat_in.cpu().numpy() if isinstance(lat_in, Tensor) else np.array(lat_in)
+        )
+        lon_in = (
+            lon_in.cpu().numpy() if isinstance(lon_in, Tensor) else np.array(lon_in)
+        )
+        lat_out = (
+            lat_out.cpu().numpy() if isinstance(lat_out, Tensor) else np.array(lat_out)
+        )
+        lon_out = (
+            lon_out.cpu().numpy() if isinstance(lon_out, Tensor) else np.array(lon_out)
+        )
+
+        (i_in, j_in) = np.mgrid[: lat_in.shape[0], : lat_in.shape[1]]
+
+        in_points = np.stack((lat_in.ravel(), lon_in.ravel()), axis=-1)
+        i_interp = LinearNDInterpolator(in_points, i_in.ravel())
+        j_interp = LinearNDInterpolator(in_points, j_in.ravel())
+
+        out_points = np.stack((lat_out.ravel(), lon_out.ravel()), axis=-1)
+        i_map = i_interp(out_points).reshape(lat_out.shape)
+        j_map = j_interp(out_points).reshape(lat_out.shape)
+
+        i_map = torch.Tensor(i_map.clip(min=0, max=lat_in.shape[0] - 2))
+        j_map = torch.Tensor(j_map.clip(min=0, max=lat_in.shape[1] - 2))
+
+        self.register_buffer("i_map", i_map)
+        self.register_buffer("j_map", j_map)
+
+    @torch.inference_mode()
+    def forward(self, values: Tensor) -> Tensor:
+        """Perform bilinear interpolation for values.
+
+        Parameters
+        ----------
+        values : torch.Tensor [..., H_in, W_in]
```

Comment:
Nit:

Move the dims ([..., H_in, W_in]) into the description line

Reply:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.